File: problems/17/17.letter-combinations-of-a-phone-number.py
#
# @lc app=leetcode id=17 lang=python3
#
# [17] Letter Combinations of a Phone Number
#


# @lc code=start
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        # digits 2-9
        # backtrack
        # f(num, i)

        if not digits:
            return []

        nums = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz",
        }
        res = []

        def backtrack(curr_combo, i):
            if i >= len(digits):
                res.append(curr_combo)
                return

            for c in nums[digits[i]]:
                backtrack(curr_combo + c, i + 1)

        backtrack("", 0)

        return res


# @lc code=end

File: problems/17/17.letter-combinations-of-a-phone-number.java
/*
 * @lc app=leetcode id=17 lang=java
 *
 * [17] Letter Combinations of a Phone Number
 */

// @lc code=start

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    private Map<Character, String> map = new HashMap<>() {
        {
            put('2', "abc");
            put('3', "def");
            put('4', "ghi");
            put('5', "jkl");
            put('6', "mno");
            put('7', "pqrs");
            put('8', "tuv");
            put('9', "wxyz");
        }
    };

    public List<String> letterCombinations(String digits) {
        /*
         * return all possible letter combinations for digits
         * 23 -> ["ad","ae","af","bd","be","bf","cd","ce","cf"]
         *
         * intuition:
         * - create hashmap of digit -> letters
         * - recursive
         *
         * O(3^N + 4^N) time, but upper bound is "9999" -> 4^4 ~ 256 times
         * O(N) space, recursive stack
         */

        List<String> res = new ArrayList<>();

        if (digits == null || digits.isEmpty()) {
            return res;
        }

        dfs(digits, 0, new StringBuilder(), res, this.map);

        return res;
    }

    private void dfs(String digits, int i, StringBuilder sb, List<String> res, Map<Character, String> map) {
        if (i == digits.length()) {
            res.add(sb.toString());
            return;
        }

        String letters = map.get(digits.charAt(i));
        for (char c : letters.toCharArray()) {
            sb.append(c);
            dfs(digits, i + 1, sb, res, map);
            // backtrack by deleting last added character
            sb.deleteCharAt(sb.length() - 1);
        }
    }
}
// @lc code=end
